sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcDeserializer.scala

```diff
@@ -73,136 +74,180 @@ class OrcDeserializer(
+      case udt: UserDefinedType[_] =>
+        val converter = newWriter(udt.sqlType)
+        (updater, ordinal, value) => {
+          converter(updater, ordinal, value)
+        }
 
       case _ =>
         throw new UnsupportedOperationException(s"$dataType is not supported yet.")
     }
 
-  private def createArrayData(elementType: DataType, length: Int): ArrayData = elementType match {
-    case BooleanType => UnsafeArrayData.fromPrimitiveArray(new Array[Boolean](length))
-    case ByteType => UnsafeArrayData.fromPrimitiveArray(new Array[Byte](length))
-    case ShortType => UnsafeArrayData.fromPrimitiveArray(new Array[Short](length))
-    case IntegerType => UnsafeArrayData.fromPrimitiveArray(new Array[Int](length))
-    case LongType => UnsafeArrayData.fromPrimitiveArray(new Array[Long](length))
-    case FloatType => UnsafeArrayData.fromPrimitiveArray(new Array[Float](length))
-    case DoubleType => UnsafeArrayData.fromPrimitiveArray(new Array[Double](length))
-    case _ => new GenericArrayData(new Array[Any](length))
+  private def getArrayDataCreator(elementType: DataType): Int => ArrayData = elementType match {
+    case BooleanType => length => UnsafeArrayData.createFreshArray(length, 1)
+    case ByteType => length => UnsafeArrayData.createFreshArray(length, 1)
+    case ShortType => length => UnsafeArrayData.createFreshArray(length, 2)
+    case IntegerType => length => UnsafeArrayData.createFreshArray(length, 4)
+    case LongType => length => UnsafeArrayData.createFreshArray(length, 8)
+    case FloatType => length => UnsafeArrayData.createFreshArray(length, 4)
+    case DoubleType => length => UnsafeArrayData.createFreshArray(length, 8)
+    case _ => length => new GenericArrayData(new Array[Any](length))
+  }
+
+  private def getRowCreator(st: StructType): () => InternalRow = {
+    val constructorsArray = new Array[() => MutableValue](st.fields.length)
+    var i = 0
+    while (i < st.fields.length) {
+      st.fields(i).dataType match {
+        case BooleanType => constructorsArray(i) = () => new MutableBoolean
+        case ByteType => constructorsArray(i) = () => new MutableByte
+        case ShortType => constructorsArray(i) = () => new MutableShort
+        // We use INT for DATE internally
+        case IntegerType | DateType => constructorsArray(i) = () => new MutableInt
+        // We use Long for Timestamp internally
+        case LongType | TimestampType => constructorsArray(i) = () => new MutableLong
+        case FloatType => constructorsArray(i) = () => new MutableFloat
+        case DoubleType => constructorsArray(i) = () => new MutableDouble
+        case _ => constructorsArray(i) = () => new MutableAny
+      }
+      i += 1
+    }
+
+    () => {
+      val array = new Array[MutableValue](constructorsArray.length)
+      var i = 0
+      while (i < constructorsArray.length) {
+        array(i) = constructorsArray(i).apply()
+        i += 1
+      }
+      new SpecificInternalRow(array)
+    }
```

Comment:
While working with avro (https://github.com/apache/spark/pull/29354) the profiler showed some time being spent in SpecificInternalRow constructor, and we saw improvements when moving to this model (please see the pdf in the avro PR) where based on the schema we can fill in a constructors array and for each data point, call these constructors one by one.

In retrospect, I did not check if the cost overhead is because of two Scala map calls in SpecificInternalRow constructors or because of a lot of case/match calls. I will investigate a bit.

Reply:
I ran the read benchmark again with changes to the SpecificInternalRow constructor to remove the map calls instead and it ran in the same time. I will update this PR and create a separate one for changes to SpecificInternalRow since it impacts both Orc and Avro (https://github.com/apache/spark/pull/29354).